src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -4479,13 +4477,18 @@ void G1CollectedHeap::eagerly_reclaim_humongous_regions() {
     }
   }
 
-  prepend_to_freelist(&local_cleanup_list);
   decrement_summary_bytes(cl.bytes_freed());
 
   phase_times()->record_fast_reclaim_humongous_time_ms((os::elapsedTime() - start_time) * 1000.0,
                                                        cl.humongous_objects_reclaimed());
 }
 
+void G1CollectedHeap::rebuild_free_region_list() {
+  Ticks start = Ticks::now();
+  _hrm.rebuild_free_list(workers());
```

Comment:
The first thing done in`rebuild_free_list()` is to drop the current free_list to rebuild if from scratch. So we should now be able to remove the call `prepend_to_freelist(&local_cleanup_list)` in `eagerly_reclaim_humongous_regions()`. Doing so the `local_cleanup_list` is only used for printing, and we should probably investigate if this can be done differently.

Reply:
There is (JDK-8264027)[https://bugs.openjdk.java.net/browse/JDK-8264027] filed for this. For now I was planning to keep this gathering of regions into the list but just factor out the printing code (as it's copy&pasted a bit), but we can certainly talk about more invasive measures.

The only way I see for now is that instead of gathering the elements in the list and printing in one go is printing immediately if we keep this functionality...

Thanks for your review.

Reply:
I agree, I think printing as we go sounds reasonable.